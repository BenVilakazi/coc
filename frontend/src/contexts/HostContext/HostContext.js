import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';
import config from '../../config';
import AlertModal from '../../components/Modal/AlertModal';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: {
    maxPlayers: config.maxPlayers.default,
    winningScore: config.winningScore.default,
    selectedPacks: config.initialSelectedPack,
    handSize: config.handSize,
    winnerScreenDisplayTime: config.winnerScreenDisplayTime,
    hideJoinCode: false,
  },
  deck: { black: [], white: [] },
  loading: [],
  czarSelection: '',
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  const [socketHasError, setSocketHasError] = useState(false);

  function createMessageHandler(currentState) {
    const czarId =
      currentState.playerIDs.length > 0 &&
      currentState.playerIDs.find(
        (playerId) => currentState.players[playerId].isCzar,
      );

    return ({ event, payload, sender }) => {
      switch (event) {
        case 'player-connected': {
          const allPlayers = [...Object.values(currentState.players)];
          const existingPlayer = allPlayers.find(
            (player) =>
              player.name.toUpperCase() === payload.playerName.toUpperCase(),
          );

          if (!existingPlayer) {
            return dispatch({ type: 'PLAYER_CONNECTED', payload });
          }

          if (existingPlayer.status === 'disconnected') {
            return dispatch({ type: 'PLAYER_RECONNECTED', payload });
          }

          return dispatch({ type: 'SEND_NAME_TAKEN_MESSAGE', payload });
        }
        case 'player-disconnected':
          return dispatch({
            type: 'PLAYER_DISCONNECTED',
            payload: {
              ...payload,
              czarId,
            },
          });

        case 'preview-winner':
          return dispatch({ type: 'PREVIEW_WINNER', payload });

        case 'winner-selected':
          return dispatch({ type: 'WINNER_SELECTED', payload });

        case 'lobby-created':
          return dispatch({ type: 'SET_LOBBY_ID', payload });

        case 'black-card-selected':
          return dispatch({ type: 'SELECT_BLACK_CARD', payload });

        case 'join-code-shuffled':
          return dispatch({ type: 'UPDATE_JOIN_CODE', payload });

        case 'player-submit':
          return dispatch({
            type: 'PLAYER_SUBMIT',
            payload: { ...payload, playerId: sender },
          });

        case 'socket-connection-error':
          dispatch({ type: 'UPDATE_JOIN_CODE', payload: { lobbyID: 'ERROR' } });
          return setSocketHasError(true);

        default:
          return undefined;
      }
    };
  }

  useEffect(() => {
    const handleMessage = createMessageHandler(state);
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, [state]);

  useEffect(() => {
    const stagedPlayerIDs = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 'staging',
    );

    const reconnectingPlayersIds = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 're-connected',
    );

    if (reconnectingPlayersIds.length > 0) {
      const reConnectingPlayers = reconnectingPlayersIds.reduce(
        (acc, id) => ({ ...acc, [id]: state.players[id] }),
        {},
      );

      dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: {
          selectedBlackCard: state.selectedBlackCard,
          players: reConnectingPlayers,
          playerIDs: reconnectingPlayersIds,
        },
      });
    }

    if (stagedPlayerIDs.length) {
      const numberOfPlayersExceedingLimit =
        state.playerIDs.length - state.gameSettings.maxPlayers;

      if (numberOfPlayersExceedingLimit > 0) {
        const playerIDsExceedingLimit = state.playerIDs.slice(
          -numberOfPlayersExceedingLimit,
        );

        dispatch({
          type: 'TOO_MANY_PLAYERS',
          payload: {
            players: playerIDsExceedingLimit,
          },
        });
      } else {
        const message =
          state.gameState === 'waiting-for-players'
            ? {
                big: "You've joined the lobby",
                small: 'Please wait for the host to start the game',
              }
            : {
                big: 'A round is in progress',
                small: 'You will join the next round automatically',
              };

        dispatch({
          type: 'SEND_PLAYER_CONNECTED_MESSAGES',
          payload: {
            players: stagedPlayerIDs,
            message,
          },
        });
      }
    }
  }, [state.playerIDs]);

  useEffect(() => {
    // if there is one player AND they are not waiting for the game to begin, end the game
    if (
      state.playerIDs.filter(
        (playerID) => state.players[playerID].status === 'playing',
      ).length === 1 &&
      state.gameState !== 'game-over'
    ) {
      dispatch({
        type: 'GAME_OVER',
        payload: { gameWinner: state.playerIDs[0], playerIDs: state.playerIDs },
      });
    }
  }, [state.playerIDs, state.players]);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
      {socketHasError && (
        <AlertModal
          bigText="SOCKET ERROR"
          smallText="Please try again later"
          buttonText="Click anywhere to restart"
          onClick={() => window.location.reload()}
        />
      )}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
