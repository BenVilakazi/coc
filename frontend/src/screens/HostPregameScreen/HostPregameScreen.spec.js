import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { HostContext } from '../../contexts/HostContext/HostContext';
import HostPregameScreen from './HostPregameScreen';

// Need to mock the Modal or createPortal errors are thrown: [Error: Target container is not a DOM element.]
// eslint-disable-next-line react/prop-types
jest.mock('../../components/Modal/Modal', () => ({ children }) => (
  <div>{children}</div>
));

jest.mock('../../components/GameSettings/GameSettings', () => () => (
  <div data-testid="game-settings" />
));

function setupFetchMock() {
  jest.spyOn(window, 'fetch').mockImplementation(() => ({
    json: async () => [
      { name: 'test 1' },
      { name: 'test 2' },
      { name: 'test 3' },
    ],
  }));
}

describe('Host Pregame Screen', () => {
  // This is the default state value provided by our context
  let state = {
    gameState: 'waiting-for-lobby',
    lobbyID: '',
    players: {},
    playerIDs: [],
    gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
    loading: [],
  };

  afterEach(() => {
    // sometimes we need to change the state for test cases, this resets to initial value
    state = {
      gameState: 'waiting-for-lobby',
      lobbyID: '',
      players: {},
      playerIDs: [],
      gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
      loading: [],
    };
    setupFetchMock();
  });

  describe('rendering', () => {
    it('renders', () => {
      const dispatch = jest.fn();
      const tree = renderer
        .create(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostPregameScreen />
          </HostContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders the loading indicator when lobbyID is an empty string', () => {
      const dispatch = jest.fn();
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('loader')).toBeInTheDocument();
    });

    it("does not render the loading indicator when lobbyID is present and loading state does not contain 'join-code'", () => {
      const dispatch = jest.fn();
      state = {
        ...state,
        lobbyID: 'ABCD',
        loading: ['aaa', 'test', 'join-cod'],
      };
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('loader')).not.toBeInTheDocument();
    });

    it("renders the loading indicator when loading state contains 'join-code'", () => {
      const dispatch = jest.fn();
      state = {
        ...state,
        loading: ['join-code'],
      };
      render(
        <HostContext.Provider
          value={{ state: { ...state, loading: ['join-code'] }, dispatch }}
        >
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(screen.getByTestId('loader')).toBeInTheDocument();
    });

    it("renders the loading indicator when loading state contains 'join-code' and a lobbyID is present", () => {
      const dispatch = jest.fn();
      state = {
        ...state,
        lobbyID: 'ABCD',
        loading: ['join-code'],
      };
      render(
        <HostContext.Provider
          value={{ state: { ...state, loading: ['join-code'] }, dispatch }}
        >
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(screen.getByTestId('loader')).toBeInTheDocument();
    });
  });

  describe('lobby', () => {
    it('creates the lobby', () => {
      const dispatch = jest.fn();
      expect(dispatch).not.toHaveBeenCalled();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch.mock.calls[0][0]).toEqual({
        type: 'CREATE_LOBBY',
        payload: {},
      });
    });
  });

  describe('buttons', () => {
    it('reloads when the close game button is clicked', async () => {
      const { reload } = window.location;

      // window.location properties are read-only, we have to redefine this object to spy on reload
      Object.defineProperty(window, 'location', {
        writable: true,
        value: { ...window.location, reload: jest.fn() },
      });

      const dispatch = jest.fn();

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      await act(async () => userEvent.click(screen.getByText('CLOSE GAME')));

      expect(window.location.reload).toHaveBeenCalledTimes(1);
      // create lobby, close game
      expect(dispatch).toHaveBeenCalledTimes(2);

      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: 'CLOSE_GAME',
        payload: {},
      });

      window.location.reload = reload;
    });

    it('calls dispatches with the proper payloads when the starting conditions are met and the start button is pressed', async () => {
      const dispatch = jest.fn();

      setupFetchMock({
        black: ['foo', 'bar', 'baz'],
        white: ['boo', 'far', 'faz'],
      });

      state = {
        gameState: 'waiting-for-lobby',
        lobbyID: '',
        players: {
          foo: {
            name: 'Bender',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          bar: {
            name: 'Briggs',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          baz: {
            name: 'Pedro',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
        },
        playerIDs: ['foo', 'bar', 'baz'],
        gameSettings: {
          maxPlayers: 8,
          winningScore: 7,
          selectedPacks: [0, 1, 2],
        },
        deck: { black: [], white: [] },
        selectedBlackCard: { text: 'test', pick: 1 },
        loading: [],
      };

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      await act(async () =>
        userEvent.click(screen.getByText('START CAROUSING')),
      );

      // create lobby, get deck, set game state, set new czar, select black card, deal white cards
      expect(dispatch).toHaveBeenCalledTimes(8);
      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: 'GET_DECK',
        payload: {},
      });
      expect(dispatch).toHaveBeenNthCalledWith(3, {
        type: 'SET_DECK',
        payload: { selectedPacks: state.gameSettings.selectedPacks },
      });
      expect(dispatch).toHaveBeenNthCalledWith(4, {
        type: 'ADD_PLAYERS_FROM_STAGING',
        payload: {},
      });
      expect(dispatch).toHaveBeenNthCalledWith(5, {
        type: 'START_GAME',
        payload: {},
      });
      expect(dispatch).toHaveBeenNthCalledWith(6, {
        type: 'SET_NEXT_CZAR',
        payload: {},
      });
      expect(dispatch).toHaveBeenNthCalledWith(7, {
        type: 'SELECT_BLACK_CARD',
        payload: {},
      });
      expect(dispatch).toHaveBeenNthCalledWith(8, {
        type: 'DEAL_WHITE_CARDS',
        payload: {},
      });
    });

    it('alerts the host if fewer than the minimum number of players are in the lobby when the start button is clicked', () => {
      const dispatch = jest.fn();

      state.playerIDs = [];
      state.players = {};

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      userEvent.click(screen.getByText('START CAROUSING'));

      // once for create lobby
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_DECK',
        payload: { selectedPacks: state.gameSettings.selectedPacks },
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'START_GAME',
        payload: {},
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_NEXT_CZAR',
        payload: {},
      });
      expect(screen.getByText('UNABLE TO START GAME')).toBeInTheDocument();
      expect(
        screen.getByText('No offense, but this game requires friends to play.'),
      ).toBeInTheDocument();
      expect(
        screen.getByText('Click anywhere to continue'),
      ).toBeInTheDocument();
    });

    it('does not allow the game to start if the amount of connected players is lower than the minimum threshold', () => {
      const dispatch = jest.fn();

      state = {
        gameState: 'waiting-for-lobby',
        lobbyID: '',
        players: {
          foo: {
            name: 'Bender',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          bar: {
            name: 'Pedro',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'disconnected',
          },
        },
        playerIDs: ['foo', 'bar'],
        gameSettings: {
          maxPlayers: 8,
          winningScore: 7,
          selectedPacks: [0, 1, 2],
        },
        deck: { black: [], white: [] },
        selectedBlackCard: { text: 'test', pick: 1 },
        loading: [],
      };

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      userEvent.click(screen.getByText('START CAROUSING'));

      // once for create lobby
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_DECK',
        payload: { selectedPacks: state.gameSettings.selectedPacks },
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'START_GAME',
        payload: {},
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_NEXT_CZAR',
        payload: {},
      });
      expect(screen.getByText('UNABLE TO START GAME')).toBeInTheDocument();
      expect(
        screen.getByText('No offense, but this game requires friends to play.'),
      ).toBeInTheDocument();
      expect(
        screen.getByText('Click anywhere to continue'),
      ).toBeInTheDocument();
    });

    it('alerts the host if no packs are selected when the start button is clicked', async () => {
      const dispatch = jest.fn();

      state = {
        gameState: 'waiting-for-lobby',
        lobbyID: '',
        players: {
          foo: {
            name: 'Bender',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          bar: {
            name: 'Briggs',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          baz: {
            name: 'Pedro',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
        },
        playerIDs: ['foo', 'bar', 'baz'],
        gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
        deck: { black: [], white: [] },
        loading: [],
      };

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      await act(async () =>
        userEvent.click(screen.getByText('START CAROUSING')),
      );

      // once for create lobby
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_DECK',
        payload: { selectedPacks: state.gameSettings.selectedPacks },
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'START_GAME',
        payload: {},
      });
      expect(dispatch).not.toHaveBeenCalledWith({
        type: 'SET_NEXT_CZAR',
        payload: {},
      });
      expect(screen.getByText('UNABLE TO START GAME')).toBeInTheDocument();
      expect(
        screen.getByText('Please pick at least one card pack.'),
      ).toBeInTheDocument();
      expect(
        screen.getByText('Click anywhere to continue'),
      ).toBeInTheDocument();
    });

    it('alerts the host if fetching cards has failed', async () => {
      // we need to mock the console.error or else jest will error out
      // https://github.com/facebook/react/issues/11098#issuecomment-523977830
      const errorSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      const dispatch = jest.fn(({ type }) => {
        // simulating a failed fetch
        if (type === 'SET_DECK') throw new Error('Failed to fetch');
      });

      state = {
        gameState: 'waiting-for-lobby',
        lobbyID: '',
        players: {
          foo: {
            name: 'Bender',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          bar: {
            name: 'Briggs',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
          baz: {
            name: 'Pedro',
            score: 0,
            isCzar: false,
            submittedCards: [],
            cards: [],
            status: 'staging',
          },
        },
        playerIDs: ['foo', 'bar', 'baz'],
        gameSettings: {
          maxPlayers: 8,
          winningScore: 7,
          selectedPacks: [0, 1, 2],
        },
        deck: { black: [], white: [] },
        loading: [],
      };

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostPregameScreen />
        </HostContext.Provider>,
      );

      await act(async () => {
        userEvent.click(screen.getByText('START CAROUSING'));
      });

      expect(screen.getByText('FAILED TO RETRIEVE CARDS')).toBeInTheDocument();
      expect(screen.getByText('Please try again later')).toBeInTheDocument();
      expect(screen.getByText('Click to restart')).toBeInTheDocument();

      errorSpy.mockRestore();
    });
  });
});
