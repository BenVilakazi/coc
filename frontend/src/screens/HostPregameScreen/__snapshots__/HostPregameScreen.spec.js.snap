// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Host Pregame Screen rendering renders 1`] = `
.c5 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c5:hover,
.c5:active,
.c5:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c1 {
  background-image: url(inactiveHamburger.svg);
  background-size: cover;
  background-color: transparent;
  border: 0;
  width: 25px;
  height: 25px;
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .hamburger-container {
  position: absolute;
  right: 17px;
  top: 11px;
  z-index: 10100;
}

.c0 .host-layout-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-layout-header h2 {
  position: relative;
  bottom: -1px;
  font-size: 1rem;
  line-height: 0.8rem;
}

.c0 .components {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c0 .components .left {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 200px;
  background-color: var(--primary-background-color);
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c0 .components .right {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
  min-width: 440px;
}

.c2 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);
  position: relative;
}

.c4 > div:not(:last-child) {
  margin-bottom: 24px;
}

.c7 .loading {
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
}

.c7 .loading div {
  display: inline-block;
  position: absolute;
  left: 8px;
  width: 16px;
  background: var(--primary-background-color);
  -webkit-animation: loading 1.2s cubic-bezier(0,0.5,0.5,1) infinite;
  animation: loading 1.2s cubic-bezier(0,0.5,0.5,1) infinite;
}

.c7 .loading div:nth-child(1) {
  left: 8px;
  -webkit-animation-delay: -0.24s;
  animation-delay: -0.24s;
}

.c7 .loading div:nth-child(2) {
  left: 32px;
  -webkit-animation-delay: -0.12s;
  animation-delay: -0.12s;
}

.c7 .loading div:nth-child(3) {
  left: 56px;
  -webkit-animation-delay: 0;
  animation-delay: 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 29px;
}

.c6 .join-code-title {
  font-size: 18px;
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  font-weight: 700;
  width: 97px;
  height: 21px;
}

.c6 .join-code {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 56px;
  font-weight: bold;
  color: var(--secondary-text-color);
  background-color: var(--secondary-background-color);
  width: 305px;
  height: 70px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  bottom: 8px;
  left: 0;
  right: 0;
}

.c10 .positionButton {
  border: none;
  background-color: transparent;
  cursor: pointer;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 4px;
  right: 4px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: var(--primary-background-color);
  padding: 8px 42px 8px 8px;
  border-radius: 5px;
  font-size: 1.8rem;
  font-weight: 700;
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c9 h2 {
  line-height: 2.4rem;
  font-size: 2rem;
}

.c9 h3 {
  font-size: 1.2rem;
  line-height: 1.2rem;
}

.c9 p {
  margin: 8px;
  font-size: 1.1rem;
}

.c9 ol {
  margin: 8px;
  line-height: 1.3rem;
}

.c9 li {
  margin: 0;
  font-size: 1rem;
  line-height: 1.3rem;
}

.c9 .howToWin {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 20px 0;
}

.c9 .snarkyText {
  font-size: 0.6rem;
  margin: 0 2px;
  padding: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c3 .player-list-wrapper {
  overflow-x: auto;
  border-bottom: 1px solid var(--primary-text-color);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 .buttons-wrapper {
  display: block;
  margin-bottom: 20px;
}

.c3 .bottom-left-wrapper {
  margin: 20px auto 32px;
  position: relative;
  bottom: 0;
  left: 0;
  right: 0;
}

.c3 .join-code-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 .host-pregame-button {
  padding: 0;
  background-color: var(--primary-background-color);
  color: var(--primary-text-color);
  margin: auto auto 12px;
}

.c3 .start-carousing {
  font-size: 0.875rem;
  padding: 16px 34px;
}

.c3 .close-game {
  font-size: 0.625rem;
  padding: 8px 20px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c8 .top {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 60%;
  overflow-y: auto;
}

.c8 .bottom {
  height: 40%;
}

@media (min-width:680px) {
  .c0 .host-layout-header {
    height: 130px;
  }

  .c0 .components .left {
    min-width: 420px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 2.5rem;
    margin-bottom: -4px;
    line-height: 2rem;
  }
}

@media (min-width:1600px) {
  .c0 .components .left {
    min-width: 700px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 3.5rem;
    margin-bottom: -35px;
    line-height: 7rem;
  }
}

@media (min-width:3200px) {
  .c0 .host-layout-header {
    height: 200px;
  }

  .c0 .components .left {
    min-width: 1100px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 6rem;
    margin-bottom: -25px;
    line-height: 7rem;
  }
}

@media (min-width:1600px) {
  .c10 .positionButton {
    margin: 0 4px;
  }

  .c10 img {
    height: 72px;
  }
}

@media (min-width:1100px) {
  .c9 {
    left: 8px;
    right: 8px;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }

  .c9 ol,
  .c9 li {
    line-height: 1.7rem;
  }
}

@media (min-width:1600px) {
  .c9 h2 {
    font-size: 3.2rem;
    line-height: 3.2rem;
  }

  .c9 h3 {
    font-size: 2.5rem;
    line-height: 2.3rem;
  }

  .c9 p {
    margin: 14px;
  }

  .c9 p,
  .c9 li {
    font-size: 1.7rem;
  }

  .c9 ol,
  .c9 li {
    line-height: 2rem;
    margin: 12px;
  }

  .c9 .snarkyText {
    font-size: 1rem;
    margin: 0 4px;
    padding: 0;
  }
}

@media (min-width:3200px) {
  .c9 h2 {
    font-size: 4rem;
    line-height: 4.4rem;
  }

  .c9 h3 {
    font-size: 3rem;
    line-height: 3rem;
  }

  .c9 p {
    margin: 20px;
    font-size: 2rem;
  }

  .c9 ol {
    margin: 40px;
  }

  .c9 li {
    margin: 10px;
    font-size: 2rem;
  }

  .c9 .howToWin {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin: 20px 0;
  }

  .c9 .snarkyText {
    font-size: 1.5rem;
    margin: 0;
    padding: 0;
  }
}

<div
  className="c0 primary-background"
>
  <div
    className="hamburger-container"
  >
    <button
      className="c1"
      onClick={[Function]}
    />
  </div>
  <div
    className="components"
  >
    <div
      className="left"
    >
      <header
        className="c2 host-layout-header"
      >
        <h2>
          CARDS OF CAROUSAL
        </h2>
      </header>
      <div
        className="c3"
      >
        <div
          className="player-list-wrapper"
        >
          <div
            className="c4"
            data-testid="playerList-container"
          />
        </div>
        <div
          className="bottom-left-wrapper"
        >
          <div
            className="buttons-wrapper"
          >
            <button
              className="c5 host-pregame-button"
              onClick={[Function]}
            >
              <div
                className="start-carousing"
              >
                START CAROUSING
              </div>
            </button>
            <button
              className="c5 host-pregame-button"
              onClick={[Function]}
            >
              <div
                className="close-game"
              >
                CLOSE GAME
              </div>
            </button>
          </div>
          <div
            className="join-code-wrapper"
          >
            <div
              className="c6"
            >
              <p
                className="join-code-title"
              >
                JOIN CODE:
              </p>
              <div
                className="join-code"
                data-testid="join-code"
              >
                <div
                  className="c7"
                  color={1}
                  data-testid="loader"
                >
                  <div
                    className="loading"
                  >
                    <div />
                    <div />
                    <div />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="right"
    >
      <div
        className="c8"
      >
        <div
          className="top"
        >
          <div
            className="c9"
          >
            <div>
              <h2>
                Welcome! ¡Bienvenidos! Добро пожаловать!
              </h2>
              <p>
                Cards of Carousal is your grandma's favorite online card game!
              </p>
              <p>
                Imagine if your indecency and crude taste in jokes was an advantage in life. It isn't, but in Cards of Carousal it can be!
              </p>
              <p>
                Long and short of this deal; one person (the Czar of the round) reads the black card on this screen. Everyone else submits a white card trying to be the Czar's favorite jester. More directions on the following pages if you're still confused.
              </p>
              <div
                className="c10"
              >
                <button
                  className="positionButton"
                  onClick={[Function]}
                  type="button"
                >
                  <img
                    alt="active position indicator"
                    src="blackCard.svg"
                  />
                </button>
                <button
                  className="positionButton"
                  onClick={[Function]}
                  type="button"
                >
                  <img
                    alt="inactive position indicator"
                    src="whiteCard.svg"
                  />
                </button>
                <button
                  className="positionButton"
                  onClick={[Function]}
                  type="button"
                >
                  <img
                    alt="inactive position indicator"
                    src="whiteCard.svg"
                  />
                </button>
                <button
                  className="positionButton"
                  onClick={[Function]}
                  type="button"
                >
                  <img
                    alt="inactive position indicator"
                    src="whiteCard.svg"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          className="bottom"
        >
          <div
            data-testid="game-settings"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
